src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs

```diff
@@ -161,17 +172,36 @@ public bool MoveNext()
 
         private unsafe void FindNextEntry()
         {
-            // Read each entry from the enumerator
-            if (Interop.Sys.ReadDir(_directoryHandle, out _entry) != 0)
+            Span<byte> buffer = _entryBuffer == null ? Span<byte>.Empty : new Span<byte>(_entryBuffer);
+            int result = Interop.Sys.ReadDir(_directoryHandle, buffer, ref _entry);
+            switch (result)
             {
-                // TODO: https://github.com/dotnet/corefx/issues/26715
-                // - Refactor ReadDir so we can process errors here
-
-                // Directory finished
-                DirectoryFinished();
+                case -1:
+                    // End of directory
+                    DirectoryFinished();
+                    break;
+                case 0:
+                    // Success
+                    break;
+                default:
+                    // Error
+                    if ((_options.IgnoreInaccessible && IsAccessError(result))
+                        || ContinueOnError(result))
+                    {
+                        DirectoryFinished();
+                        break;
+                    }
+                    else
+                    {
+                        throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(result), _currentPath, isDirectory: true);
+                    }
             }
         }
 
+        private static bool IsAccessError(int error)
```

Comment:
Worth moving to Interop.IOErrors.cs ?

Reply:
Possibly. I'll look at that when I do the directory handle stuff. We'd want to do the same for Windows if we did.